### PR TITLE
Add custom script engines for donwloading/uplaod/listing

### PR DESCRIPTION
**Pull request by tmspzz**
See #146.

It _should_ also work for:  #120, #139, #116

### CODE REVIEW

**Comment by grioja** [comment]
Any word on this PR? 👍 for getting it in. We need to add caching support, but have on our on-site file store that we would like to use instead of S3.

**Comment by blender** [comment]
@grioja I work on this on and off, contributions are welcome.

In the mean time, setting up minio https://www.minio.io/ might get you around the problem.

File: src/Lib.hs [hunk 1]
```diff
@@ -777,6 +836,56 @@ saveFrameworkAndArtifactsToLocalCache lCacheDir reverseRomeMap fVersion@(Framewo
...
+    --   frameworkArchive <- createZipArchive frameworkDirectory verbose
+  where
+    frameworkNameWithFrameworkExtension = appendFrameworkExtensionTo f
+    platformBuildDirectory =
+      carthageArtifactsBuildDirectoryForPlatform platform f
+    frameworkDirectory =
+      platformBuildDirectory </> frameworkNameWithFrameworkExtension
+    dSYMNameWithDSYMExtension = frameworkNameWithFrameworkExtension <> ".dSYM"
+    dSYMdirectory = platformBuildDirectory </> dSYMNameWithDSYMExtension
+    bcSybolMapPath d = platformBuildDirectory </> bcsymbolmapNameFrom d
```
Comment (fabb) [inline src/Lib.hs]: Typo

**Comment by BalestraPatrick** [comment]
@blender @fabb What's missing to land this PR? I _may_ try to help out a little bit with my very little knowledge of Haskell if something is needed.

**Comment by blender** [comment]
@BalestraPatrick basically the path is laid out. What is missing is the implementation.

Whenever you read `undefined`, that is where the gap needs filling.

**Comment by blender** [comment]
I have resolved the conflicts are rebased on #179

**Comment by BalestraPatrick** [comment]
Can you elaborate a little bit on the idea for your implementation?

**Comment by blender** [comment]
@BalestraPatrick the idea is to invoke a script at `mEnginePath` and pass it a bunch on arguments:

First argument is always the command currently being executed: i.e. `download`, `upload` or `list`

In case of `download`, arguments are:
1) `download`
2) `remote-path-to-the-file-as-constructed-by-rome`
3) `local-path-as-constructed-by-rome`

In case of `upload`, arguments are:
1) `upload`
2) `local-path-as-constructed-by-rome`
3) `remote-path-to-the-file-as-constructed-by-rome`

In case of `list`, arguments are:
1) `list`
2) `remote-path-to-the-file-as-constructed-by-rome`

scripts should return exit code 0 on success and be passed stdin, stdout and stderr descriptors by rome itself. You can use [Turtle](https://hackage.haskell.org/package/turtle) to invoke the scripts.

**Comment by blender** [comment]
This is to start with so that things don't get too complicated. 

Alternatively the idea was to pass a description of the data structure that the commands rely upon to the script. However I don't have a concise solution for this. One idea is to use JSON to describe them but I think for a first implementation this is an overkill.

**Pull request closed without merging**